core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -642,16 +672,26 @@ class Partition(val topicPartition: TopicPartition,
    *                   topic partition
    * @param isr The set of broker ids that are known to be insync with the leader
    */
-  def updateAssignmentAndIsr(assignment: Seq[Int], isr: Set[Int]): Unit = {
+  def updateAssignmentAndIsr(assignment: Seq[Int],
+                             isr: Set[Int],
+                             addingReplicas: Seq[Int] = Seq.empty,
+                             removingReplicas: Seq[Int] = Seq.empty): Unit = {
     val replicaSet = assignment.toSet
     val removedReplicas = remoteReplicasMap.keys -- replicaSet
 
+    // if we're already in a reassignment then let's carry the original replicas
```

Comment:
I think the KIP was not very clear about this, but we shouldn't need to rely on the previous state to compute the original assignment. Basically what I had in mind is this:

```
sizeOf(original assignment) == sizeOf(replicas - adding replicas) 
```

For example, consider a reassignment from [1, 2, 3] to [4, 5, 6]. When the reassignment kicks off, assignment will be equal to [1, 2, 3, 4, 5, 6], with [4, 5, 6] as adding replicas. So original assignment is computed as [1, 2, 3, 4, 5, 6] - [4, 5, 6]. Then even if the ISR happens to be [1, 4, 5], then we do not count it as an URP.

The reason to consider only the original assignment is that we do not want URPs when when increasing the replication factor. Reassignment completion depends on reaching the new replication factor, after which the AddingReplica set will go back to empty and the URP calculation will take into account the new replication factor.

Reply:
Yep, it's a good idea that we can compute it on the fly. I'll update the code, thanks.